services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/muutosilmoitus/MuutosilmoitusService.kt

```diff
@@ -51,6 +61,149 @@ class MuutosilmoitusService(
         return saved
     }
 
+    @Transactional
+    fun update(id: UUID, request: HakemusUpdateRequest, currentUserId: String): Muutosilmoitus {
+        logger.info("Updating muutosilmoitus id=$id")
+
+        val entity =
+            muutosilmoitusRepository.findByIdOrNull(id) ?: throw MuutosilmoitusNotFoundException(id)
+        if (entity.sent != null) throw MuutosilmoitusAlreadySentException(entity)
+        val originalMuutosilmoitus = entity.toDomain()
+
+        logger.info { "The muutosilmoitus to update is ${entity.logString()}" }
+
+        assertUpdateCompatible(entity, request)
+
+        if (!request.hasChanges(originalMuutosilmoitus.hakemusData)) {
+            logger.info("Not updating unchanged hakemus data. ${entity.id}")
```

Comment:
Should this say "... unchanged muutosilmoitus data."?

Reply:
I think this is correct as-is. It's the hakemus data that hasn't changed.